### PR TITLE
[OPT] Run test in lower precision on GPU

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

Let's run the test only in half precision. There might have been a bug because fp16 weights are loaded into fp32. Let's make sure everything stays in fp16.

Also if it fails again, we'll see a better error message this time.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by ArthurZucker — APPROVED** [review]

File: src/transformers/models/opt/modeling_opt.py [hunk 1]
```diff
@@ -86,52 +85,28 @@ def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int]
...
-        return F.embedding(
-            positions,
-            self.weight,
-            self.padding_idx,
-            self.max_norm,
-            self.norm_type,
-            self.scale_grad_by_freq,
-            self.sparse,
-        )
+    def __init__(self, num_embeddings: int, embedding_dim: int):
```
Comment (patrickvonplaten) [inline src/transformers/models/opt/modeling_opt.py:93]: Cleaned this up a bit. Think we should try to align it as much as possible to Bart here

File: tests/models/opt/test_modeling_opt.py [hunk 2]
```diff
@@ -268,17 +268,22 @@ def _long_tensor(tok_lst):
 @require_torch
 class OPTModelIntegrationTests(unittest.TestCase):
     @slow
+    @unittest.skipIf(torch_device != "cpu", "Cannot make deterministic on GPU")
```
Comment (patrickvonplaten) [inline tests/models/opt/test_modeling_opt.py:271]: Getting different logits results on GPU depending on PyTorch version (1.10+cu11.0 vs. 1.11+cu11.4) and results also differ between CPU and GPU. Only on CPU it seems to be deterministic.

It's not because the weights are saved & loaded in FP16 - checked that the same happens when weights are stored in fp32 and loaded in fp32. 
It's across multiple checkpoints (not just 350m, but also 125m).
The differences start to creep in in the first linear projection matrix `project_in_dim` for the 350m model and the first attention layer for 125m.

Comment (patrickvonplaten) [reply tests/models/opt/test_modeling_opt.py]: Upgrading precision to 1e-2 so that tests pass on GPU - think that's the best we can do

Comment (patrickvonplaten) [reply tests/models/opt/test_modeling_opt.py]: **Interesting observation:**

- BART has the same problem, but the differences are not as flagrant in the end
- BERT does **not** have this problem at all. Results are exactly the same.

cc @stephenroller @stas00 @sgugger @LysandreJik

File: src/transformers/models/opt/modeling_opt.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/opt/modeling_opt.py:93]: Thanks :) , I have to update the tf code based on that I think

File: tests/models/opt/test_modeling_opt.py [hunk 3]
```diff
@@ -269,16 +269,28 @@ def _long_tensor(tok_lst):
 class OPTModelIntegrationTests(unittest.TestCase):
     @slow
     def test_inference_no_head(self):
+        # model is not deterministic on GPU, not sure why
```
Comment (ydshieh) [inline tests/models/opt/test_modeling_opt.py:272]: (some) non-deterministic is mentioned
https://pytorch.org/docs/stable/notes/randomness.html

The diff `1e-2` is still unpleasantly large though.

**Review by ydshieh — APPROVED** [review]
I will allow `1e-2` for now 😄

File: tests/models/opt/test_modeling_opt.py (hunk 3 above)
Comment (patrickvonplaten) [reply tests/models/opt/test_modeling_opt.py:272]: Actually my comment is not great - the model is deterministic on GPU, but it's just that results differ depending on the CUDA vesion and CUDA != CPU

Comment (patrickvonplaten) [inline tests/models/opt/test_modeling_opt.py:272]:
```suggestion
```

File: tests/models/opt/test_modeling_opt.py (hunk 2 above)
Comment (stas00) [reply tests/models/opt/test_modeling_opt.py]: getting logits to exact match is very difficult. Perhaps a much longer input and checking that softmax output matches - after all logits fluctuations don't matter at the end but 100% softmax match does.

Comment (stas00) [reply tests/models/opt/test_modeling_opt.py]: if you use a very high tolerance then it becomes questionable whether the test is doing anything.

**Review by sgugger — APPROVED** [review]
Thanks for fixing!

File: tests/models/opt/test_modeling_opt.py (hunk 3 above)
Comment (ydshieh) [reply tests/models/opt/test_modeling_opt.py]: @NielsRogge has this problem too before, and we end up with 2 expected values, one for CPU and one for GPU.
But now you mentioned different CUDA versions also count, so that approach is not super scalable.
